classes/task/resendadhoc_task.php

```diff
@@ -0,0 +1,159 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace logstore_xapi\task;
+
+require_once($CFG->dirroot . '/admin/tool/log/store/xapi/lib.php');
+
+/**
+ * Schedules a adhoc task resendadhoc.
+ *
+ * @package   logstore_xapi
+ * @copyright Scott Verbeek <scottverbeek@catalyst-au.net>
+ * @license   https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class resendadhoc_task extends \core\task\adhoc_task {
+
+    public static function instance(
+        string $errortype,
+        string $eventname,
+        int $datefrom,
+        int $dateto,
+        int $runtime,
+        int $batch
+    ): self {
+        $task = new self();
+        $task->set_custom_data((object) [
+            'errortype' => $errortype,
+            'eventname' => $eventname,
+            'datefrom' => $datefrom,
+            'dateto' => $dateto,
+            'runtime' => $runtime,
+            'batch' => $batch
+        ]);
+
+        return $task;
+    }
+
+    /**
+     * Do the job.
+     * Throw exceptions on errors (the job will be retried).
+     */
+    public function execute() {
+        global $DB;
+
+        $options = $this->get_custom_data();
+
+        $basetable = XAPI_REPORT_SOURCE_FAILED;
+        $params = [];
+        $where = [];
+
+        // Start of sanitization, and applying of scope.
+        mtrace("Program will stop after {$options->runtime} seconds have passed, started at " .date('Y-m-d H:i:s T') . "...");
```

Comment:
Might be worth changing messages that say "program" to something like "task" or "adhoc task"